File: crypto/src/test/cplusplus/eqextension.cpp
/*
 * Copyright (c) 2024-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <boost/test/unit_test.hpp>
#include <ranges>

#include "circuitbuilder.h"
#include "customizableconstraintsystem.h"
#include "eqextension.h"
#include "hypercube.h"
#include "point.h"
#include "solinas62.h"
#include "util.h"

using namespace blacknet::crypto;

using E = Solinas62Ring;

BOOST_AUTO_TEST_SUITE(EqExtension_Plain)

BOOST_AUTO_TEST_CASE(meta) {
    std::vector<E> a{E(1), E(0), E(0)};
    EqExtension eq(a);
    BOOST_TEST(1 == eq.degree());
    BOOST_TEST(3 == eq.variables());
}

BOOST_AUTO_TEST_CASE(mul) {
    EqExtension<E> a({E(2), E(3), E(5), E(7)});
    E b(11);
    Point<E> r{E(13), E(17), E(23), E(27)};
    BOOST_TEST(a(r) * b == (a * b)(r));
}

BOOST_AUTO_TEST_CASE(bind) {
    EqExtension<E> eq1({E(2), E(3), E(4)});
    Point<E> r1{E(5), E(6), E(7)};
    EqExtension<E> eq2(eq1);
    eq2.bind(E(5));
    Point<E> r2{E(6), E(7)};
    EqExtension<E> eq3(eq2);
    eq3.bind(E(6));
    Point<E> r3{E(7)};
    BOOST_TEST(eq1(r1) == eq2(r2));
    BOOST_TEST(eq1(r1) == eq3(r3));

    std::vector<E> evaluations(4);
    EqExtension<E> eq = eq1;
    eq.bind(E(-2));
    eq1.bind<E(-2), util::Assign<E>>(evaluations);
    BOOST_TEST(eq() == evaluations);
    eq = eq1;
    eq.bind(E(-1));
    eq1.bind<E(-1), util::Assign<E>>(evaluations);
    BOOST_TEST(eq() == evaluations);
    eq = eq1;
    eq.bind(E(0));
    eq1.bind<E(0), util::Assign<E>>(evaluations);
    BOOST_TEST(eq() == evaluations);
    eq = eq1;
    eq.bind(E(1));
    eq1.bind<E(1), util::Assign<E>>(evaluations);
    BOOST_TEST(eq() == evaluations);
    eq = eq1;
    eq.bind(E(2));
    eq1.bind<E(2), util::Assign<E>>(evaluations);
    BOOST_TEST(eq() == evaluations);
    eq = eq1;
    eq.bind(E(3));
    eq1.bind<E(3), util::Assign<E>>(evaluations);
    BOOST_TEST(eq() == evaluations);
    eq = eq1;
    eq.bind(E(4));
    eq1.bind<E(4), util::Assign<E>>(evaluations);
    BOOST_TEST(eq() == evaluations);

    std::vector<E> pis(eq2());
    Hypercube<E> hc(eq2.variables());
    for (const auto& [index, b] : std::views::zip(
            std::ranges::subrange(hc.composedBegin(), hc.composedEnd()),
            std::ranges::subrange(hc.decomposedBegin(), hc.decomposedEnd())
        )) {
        BOOST_TEST(eq2(b) == pis[index]);
    };
}

BOOST_AUTO_TEST_CASE(point) {
    Hypercube<E> hc(3);
    Point<E> a{E(1), E(0), E(0)};
    EqExtension eq(a.coordinates);
    std::ranges::for_each(hc.decomposedBegin(), hc.decomposedEnd(), [&](const Point<E>& b) {
        if (a == b)
            BOOST_TEST(E(1) == eq(b));
        else
            BOOST_TEST(E(0) == eq(b));
    });
}

BOOST_AUTO_TEST_CASE(hypercube) {
    Hypercube<E> hc(3);
    std::vector<E> a{E(1), E(0), E(0)};
    EqExtension eq(a);
    std::vector<E> pis(eq());
    std::ranges::for_each(hc.composedBegin(), hc.composedEnd(), [&](const std::size_t& i) {
        if (i == 4)
            BOOST_TEST(E(1) == pis[i]);
        else
            BOOST_TEST(E(0) == pis[i]);
    });
}

BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE(EqExtension_Circuit)

BOOST_AUTO_TEST_CASE(points) {
    EqExtension<E> eq({E(2), E(3), E(5)});
    Point<E> x{E(7), E(11), E(13)};

    using Builder = CircuitBuilder<E, 2>;
    Builder circuit;
    using EqCircuit = EqExtension<E>::Circuit<Builder>;
    EqCircuit eq_circuit(circuit, Builder::Variable::Type::Input, 3);
    using PointCircuit = Point<E>::Circuit<Builder>;
    PointCircuit x_circuit(circuit, Builder::Variable::Type::Input, 3);
    eq_circuit(x_circuit);

    CustomizableConstraintSystem<E> ccs(circuit.ccs());
    VectorDense<E> z = ccs.assigment();
    std::ranges::copy(eq.coefficients, std::back_inserter(z.elements));
    std::ranges::copy(x, std::back_inserter(z.elements));

    using EqAssigner = EqExtension<E>::Assigner<Builder::degree()>;
    EqAssigner eq_assigner(z.elements);
    BOOST_TEST(eq(x) == eq_assigner.point(eq, x));
    BOOST_TEST(ccs.isSatisfied(z));
}

BOOST_AUTO_TEST_CASE(hypercubes) {
    EqExtension<E> eq({E(29), E(31), E(37)});

    using Builder = CircuitBuilder<E, 2>;
    Builder circuit;
    using EqCircuit = EqExtension<E>::Circuit<Builder>;
    EqCircuit eq_circuit(circuit, Builder::Variable::Type::Input, 3);
    EqCircuit::hypercube(circuit, eq_circuit.coefficients);

    CustomizableConstraintSystem<E> ccs(circuit.ccs());
    VectorDense<E> z = ccs.assigment();
    std::ranges::copy(eq.coefficients, std::back_inserter(z.elements));

    using EqAssigner = EqExtension<E>::Assigner<Builder::degree()>;
    EqAssigner eq_assigner(z.elements);
    BOOST_TEST(eq() == eq_assigner.hypercube(eq.coefficients));
    BOOST_TEST(ccs.isSatisfied(z));
}

BOOST_AUTO_TEST_SUITE_END()

File: crypto/src/test/rust/eqextension.rs
/*
 * Copyright (c) 2024-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use blacknet_compat::assert_ok;
use blacknet_crypto::assigner::eqextension::EqExtension as Assigner;
use blacknet_crypto::circuit::circuitbuilder::{CircuitBuilder, VariableKind};
use blacknet_crypto::circuit::eqextension::EqExtension as Circuit;
use blacknet_crypto::circuit::point::Point as PointCircuit;
use blacknet_crypto::constraintsystem::ConstraintSystem;
use blacknet_crypto::eqextension::EqExtension;
use blacknet_crypto::hypercube::Hypercube;
use blacknet_crypto::point::Point;
use blacknet_crypto::polynomial::Polynomial;
use blacknet_crypto::ring::{Ring, UnitalRing};

type R = blacknet_crypto::pervushin::PervushinField;

#[test]
fn meta() {
    let eq = EqExtension::from([1, 0, 0].map(R::from));
    assert_eq!(eq.degree(), 1);
    assert_eq!(eq.variables(), 3);
}

#[test]
fn mul() {
    let a = EqExtension::from([2, 3, 5, 7].map(R::from));
    let b = R::from(11);
    let r = Point::from([13, 17, 23, 27].map(R::from));
    assert_eq!(a.point(&r) * b, (a * b).point(&r));
}

#[test]
fn bind() {
    let eq1 = EqExtension::from([2, 3, 4].map(R::from));
    let r1 = Point::from([5, 6, 7].map(R::from));
    let mut eq2 = eq1.clone();
    eq2.bind(R::from(5));
    let r2 = Point::from([6, 7].map(R::from));
    let mut eq3 = eq2.clone();
    eq3.bind(R::from(6));
    let r3 = Point::from([7].map(R::from));
    assert_eq!(eq2.point(&r2), eq1.point(&r1));
    assert_eq!(eq3.point(&r3), eq1.point(&r1));

    let mut eq = eq1.clone();
    eq.bind(R::from(-2));
    let evaluations = eq1.hypercube_with_var::<-2>();
    assert_eq!(evaluations, eq.hypercube());
    eq = eq1.clone();
    eq.bind(R::from(-1));
    let evaluations = eq1.hypercube_with_var::<-1>();
    assert_eq!(evaluations, eq.hypercube());
    eq = eq1.clone();
    eq.bind(R::from(0));
    let evaluations = eq1.hypercube_with_var::<0>();
    assert_eq!(evaluations, eq.hypercube());
    eq = eq1.clone();
    eq.bind(R::from(1));
    let evaluations = eq1.hypercube_with_var::<1>();
    assert_eq!(evaluations, eq.hypercube());
    eq = eq1.clone();
    eq.bind(R::from(2));
    let evaluations = eq1.hypercube_with_var::<2>();
    assert_eq!(evaluations, eq.hypercube());
    eq = eq1.clone();
    eq.bind(R::from(3));
    let evaluations = eq1.hypercube_with_var::<3>();
    assert_eq!(evaluations, eq.hypercube());
    eq = eq1.clone();
    eq.bind(R::from(4));
    let evaluations = eq1.hypercube_with_var::<4>();
    assert_eq!(evaluations, eq.hypercube());
}

#[test]
fn point() {
    let hc = Hypercube::<R>::new(3);
    let a = Point::<R>::from([1, 0, 0].map(R::from));
    let eq = EqExtension::<R>::from(a.coordinates().clone());
    hc.iter_vertex().for_each(|b| {
        if a == b {
            assert_eq!(eq.point(&b), R::UNITY);
        } else {
            assert_eq!(eq.point(&b), R::ZERO);
        }
    });
}

#[test]
fn hypercube() {
    let hc = Hypercube::<R>::new(3);
    let a = Vec::<R>::from([1, 0, 0].map(R::from));
    let eq = EqExtension::<R>::from(a);
    let pis = eq.hypercube();
    hc.iter_index().for_each(|i| {
        if i == 4 {
            assert_eq!(pis[i], R::UNITY);
        } else {
            assert_eq!(pis[i], R::ZERO);
        }
    });
}

#[test]
fn circuit_point() {
    let coefficients_plain = [2, 3, 5].map(R::from);
    let eq_plain = EqExtension::<R>::from(coefficients_plain);
    let x_plain = Point::<R>::from([7, 11, 13].map(R::from));
    let y_plain = eq_plain.point(&x_plain);

    let circuit = CircuitBuilder::<R>::new(2);
    let scope = circuit.scope("test");
    let eq_circuit = Circuit::allocate(&circuit, VariableKind::PublicInput, eq_plain.variables());
    let x_circuit =
        PointCircuit::allocate(&circuit, VariableKind::PublicInput, x_plain.dimension());
    let _y_circuit = eq_circuit.point(&x_circuit);
    drop(scope);

    let r1cs = circuit.r1cs();
    let z = r1cs.assigment();
    z.extend(coefficients_plain);
    z.extend_from_slice(&x_plain);

    let eq_assigner = Assigner::new(coefficients_plain.into(), &z);
    let y_assigned = eq_assigner.point(&x_plain);

    assert_eq!(y_assigned, y_plain);
    assert_ok!(r1cs.is_satisfied(&z.finish()));
}

#[test]
fn circuit_hypercube() {
    let coefficients_plain = [29, 31, 37].map(R::from);
    let eq_plain = EqExtension::<R>::from(coefficients_plain);
    let y_plain = eq_plain.hypercube();

    let circuit = CircuitBuilder::<R>::new(2);
    let scope = circuit.scope("test");
    let eq_circuit = Circuit::allocate(&circuit, VariableKind::PublicInput, eq_plain.variables());
    let _y_circuit = eq_circuit.hypercube();
    drop(scope);

    let r1cs = circuit.r1cs();
    let z = r1cs.assigment();
    z.extend(coefficients_plain);

    let eq_assigner = Assigner::new(coefficients_plain.into(), &z);
    let y_assigned = eq_assigner.hypercube();

    assert_eq!(y_assigned, y_plain);
    assert_ok!(r1cs.is_satisfied(&z.finish()));
}
